lib/class-wp-rest-image-editor-controller.php

```diff
@@ -101,90 +101,134 @@ public function permission_callback( $request ) {
-		if ( ! $image_editor->load() ) {
-			return new WP_Error( 'rest_cannot_load_editor', __( 'Unable to load original media file.', 'gutenberg' ), array( 'status' => 500 ) );
+			// TODO: return the "real" error message?
+			if ( is_wp_error( $result ) ) {
+				return new WP_Error( 'rest_image_rotation_failed', __( 'Unable to rotate this image.', 'gutenberg' ), array( 'status' => 500 ) );
+			}
 		}
 
-		if ( isset( $params['rotation'] ) ) {
-			$image_editor->rotate( 0 - $params['rotation'] );
+		if ( isset( $params['x'], $params['y'], $params['width'], $params['height'] ) ) {
+			$size = $image_editor->get_size();
+
+			// Crop
+			$crop_x = round( ( $size['width'] * floatval( $params['x'] ) ) / 100.0 );
+			$crop_y = round( ( $size['height'] * floatval( $params['y'] ) ) / 100.0 );
+			$width  = round( ( $size['width'] * floatval( $params['width'] ) ) / 100.0 );
+			$height = round( ( $size['height'] * floatval( $params['height'] ) ) / 100.0 );
+
+			$result = $image_editor->crop( $crop_x, $crop_y, $width, $height );
+
+			if ( is_wp_error( $result ) ) {
+				return new WP_Error( 'rest_image_crop_failed', __( 'Unable to crop this image.', 'gutenberg' ), array( 'status' => 500 ) );
+			}
 		}
 
-		$size = $image_editor->get_size();
+		// Calculate the file name.
+		$image_ext  = pathinfo( $image_file, PATHINFO_EXTENSION );
+		$image_name = wp_basename( $image_file, ".{$image_ext}" );
+
+		// Do not append multiple `-edited` to the file name.
+		// The user may be editing a previously edited image.
+		if ( preg_match( '/-edited(-\d+)?$/', $image_name ) ) {
+			// Remove any `-1`, `-2`, etc. `wp_unique_filename()` will add the proper number.
+			$image_name = preg_replace( '/-edited(-\d+)?$/', '-edited', $image_name );
+		} else {
+			// Append `-edited` before the extension.
+			$image_name .= '-edited';
+		}
 
-		// Finally apply the modifications.
-		$crop_x = round( ( $size['width'] * floatval( $params['x'] ) ) / 100.0 );
-		$crop_y = round( ( $size['height'] * floatval( $params['y'] ) ) / 100.0 );
-		$width  = round( ( $size['width'] * floatval( $params['width'] ) ) / 100.0 );
-		$height = round( ( $size['height'] * floatval( $params['height'] ) ) / 100.0 );
-		$image_editor->crop( $crop_x, $crop_y, $width, $height );
+		$filename = "{$image_name}.{$image_ext}";
 
-		// TODO: Generate filename based on edits.
-		$target_file = 'edited-' . $meta['original_name'];
+		// Create the uploads sub-directory if needed.
+		$uploads = wp_upload_dir();
 
-		$filename = rtrim( dirname( $image_path ), '/' ) . '/' . $target_file;
+		// Tests the file basename in the (proper/new) upload directory.
+		$filename = wp_unique_filename( $uploads['path'], $filename );
 
 		// Save to disk.
-		$saved = $image_editor->save( $filename );
+		$saved = $image_editor->save( $uploads['path'] . "/$filename" );
 
 		if ( is_wp_error( $saved ) ) {
 			return $saved;
 		}
 
-		// Update attachment details.
+		// Create new attachment post.
```

Comment:
Should we copy the rest of the attachment data?

Reply:
Yeah, may make sense. But then we'll get two (or more) attachments with the same description and caption. In general don't think the image alt text should be carried over as alt has to be contextual. Editing an older, already used image would suggest it is being used in different context. Editing an image just after uploading it won't have alt text.